core/src/apps/cardano/sign_tx.py

```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
 class Transaction:
     def __init__(
         self,
-        inputs: list,
-        outputs: list,
-        keychain,
+        inputs: List[CardanoTxInputType],
+        outputs: List[CardanoTxOutputType],
+        keychain: seed.Keychain,
         protocol_magic: int,
-        input_coins_sum: int,
-    ):
+        fee: int,
+        ttl: int,
+    ) -> None:
         self.inputs = inputs
         self.outputs = outputs
         self.keychain = keychain
-        # attributes have to be always empty in current Cardano
-        self.attributes = {}
+        self.fee = fee
+        self.ttl = ttl
 
         self.network_name = KNOWN_PROTOCOL_MAGICS.get(protocol_magic, "Unknown")
         self.protocol_magic = protocol_magic
-        self.input_coins_sum = input_coins_sum
 
-    def _process_outputs(self):
-        change_addresses = []
-        change_derivation_paths = []
-        output_addresses = []
-        outgoing_coins = []
-        change_coins = []
+    def _build_inputs(self) -> list:
+        inputs_for_cbor = []
+        for input in self.inputs:
+            inputs_for_cbor.append([input.prev_hash, input.prev_index])
```

Comment:
IMHO this deserves to be a tupple and there should be no conflict in encoding as I see that the cbor encoder does not have a rule for encoding tuples yet: https://github.com/vacuumlabs/trezor-firmware/blob/ca3dd2bfb9b02cfad8077f77e680740e6de3d9e0/core/src/apps/common/cbor.py#L70 so I would extend it to support them. 

This will allow you to put here a more specific return type which increases readability

Reply:
same for the remaining `_build` functions

Reply:
regarding the CBOR rule for tuples, that'sin hindsight probably too radical as this may complicate the decoding logic as there would not be 1:1 mapping as the same coding would be used for lists as for tuples. Still, you can return tuples from there and cast them to lists just before cbor-encoding

Reply:
Yes, I though about that, I just didn't want to do `list(returned_value)` each time. I'll do it then.

Reply:
Actually no. I don't think it'll work. I return lists from those functions. And those lists are then directly added to tx_body. So I would have iterate through them again. ... We'll discuss it offline.

Reply:
In the end we decided to return lists of tuples from these functions. We then use a `_detupleize` function to convert those lists of tuples into lists of lists (as required by CBOR).